src/pykeen/sampling/basic_negative_sampler.py

```diff
@@ -2,60 +2,122 @@
+    def __init__(
+        self,
+        triples_factory: TriplesFactory,
+        num_negs_per_pos: Optional[int] = None,
+        filtered: bool = False,
+        corruption_scheme: Optional[Collection[str]] = None,
+    ) -> None:
+        """Initialize the negative sampler with the given entities.
+
+        :param triples_factory: The factory holding the triples to sample from
+        :param num_negs_per_pos: Number of negative samples to make per positive triple. Defaults to 1.
+        :param filtered: Whether proposed corrupted triples that are in the training data should be filtered.
+            Defaults to False.
+        :param corruption_scheme: What sides ('h', 'r', 't') should be corrupted. Defaults to head and tail ('h', 't').
+        """
+        super().__init__(
+            triples_factory=triples_factory,
+            num_negs_per_pos=num_negs_per_pos,
+            filtered=filtered,
+        )
+        self.corruption_scheme = corruption_scheme or ('h', 't')
+        # Set the indices
+        self._corruption_indices = [LOOKUP[side] for side in self.corruption_scheme]
+
+    def sample(self, positive_batch: torch.LongTensor) -> Tuple[torch.LongTensor, Optional[torch.Tensor]]:
         """Generate negative samples from the positive batch."""
         if self.num_negs_per_pos > 1:
             positive_batch = positive_batch.repeat(self.num_negs_per_pos, 1)
 
         # Bind number of negatives to sample
         num_negs = positive_batch.shape[0]
 
-        # Equally corrupt head and tail
-        split_idx = num_negs // 2
+        # Equally corrupt all sides
+        split_idx = num_negs // len(self._corruption_indices)
 
         # Copy positive batch for corruption.
         # Do not detach, as no gradients should flow into the indices.
         negative_batch = positive_batch.clone()
 
         # Sample random entities as replacement
-        negative_entities = torch.randint(high=self.num_entities - 1, size=(num_negs,), device=positive_batch.device)
-
-        # Replace heads – To make sure we don't replace the head by the original value
-        # we shift all values greater or equal than the original value by one up
-        # for that reason we choose the random value from [0, num_entities -1]
-        filter_same_head = (negative_entities[:split_idx] >= positive_batch[:split_idx, 0])
-        negative_batch[:split_idx, 0] = negative_entities[:split_idx] + filter_same_head.long()
-        # Corrupt tails
-        filter_same_tail = (negative_entities[split_idx:] >= positive_batch[split_idx:, 2])
-        negative_batch[split_idx:, 2] = negative_entities[split_idx:] + filter_same_tail.long()
-
-        return negative_batch
+        if 0 in self._corruption_indices or 2 in self._corruption_indices:
+            negative_entities = torch.randint(
+                high=self.num_entities - 1,
+                size=(num_negs,),
+                device=positive_batch.device,
+            )
```

Comment:
If we corrupt an entity position *and* a relation we generate too many negative integers. Maybe we can just move the generation in the 
```python
for index, start in zip(self._corruption_indices, range(0, num_negs, split_idx)):
```
loop?

Reply:
Then we also do not need the indexing on the rhs
```python
 negative_relations[start:stop]
```

Reply:
I considered creating it dynamically within the loop. I just tested it speed-wise and it looks like it definitely doesn't have adverse effects. But we need a distinction between relations and entities, because the numbers that they each base on `high=self.num_relations - 1,` and `high=self.num_entities - 1,` respectively.

Reply:
Doesn't get cleaner than this https://github.com/pykeen/pykeen/pull/209/commits/a0766fecb662af1b1701f73dde1d930b9a228aa0 😄